exercises/dominoes/.meta/src/reference/java/Dominoes.java

```diff
@@ -0,0 +1,69 @@
+import java.util.ArrayList;
+import java.util.Arrays;
+
+class Dominoes {
+	
+	ArrayList<Domino> formChain(ArrayList<Domino> inputDominoes) {
+		if (inputDominoes.size() == 0) {
+			throw new IllegalArgumentException("Empty input domino list.");
+		} else {
+			ArrayList<Domino> dominoChain = formPartialChain(new ArrayList<Domino>(), inputDominoes);
+			if(dominoChain == null) {
+				throw new IllegalArgumentException("No domino chain found.");
+			} else {
+				return dominoChain;
+			}
+		}
+	}
+	
+	private boolean isValidChain(ArrayList<Domino> chain) {
+		return isValidPartialChain(chain) && (chain.get(0).getLeft() == chain.get(chain.size() - 1).getRight());
+	}
+	
+	private boolean isValidPartialChain(ArrayList<Domino> chain) {
+		for (int i = 0; i < chain.size() - 1; i++) {
+			if (chain.get(i).getRight() != chain.get(i+1).getLeft()) {
+				return false;
+			}
+		}
+		
+		return true;
+	}
+	
+	ArrayList<Domino> formPartialChain(ArrayList<Domino> current, ArrayList<Domino> remaining) {
+		if (remaining.size() == 0) {
+			if (isValidChain(current)) {
+				return current;
+			} else {
+				return null;
```

Comment:
Even in internal methods I would prefer to throw exceptions over returning null. That makes the code neater and it means people maintaining your code don't have to remember to check for null 🙂

Reply:
:+1: Exceptions over null!! :fireworks: